nifi-nar-bundles/nifi-mqtt-bundle/nifi-mqtt-processors/src/main/java/org/apache/nifi/processors/mqtt/ConsumeMQTT.java

```diff
@@ -0,0 +1,341 @@
+                mqttQueue.drainTo(newBuffer);
+                mqttQueue = newBuffer;
+            }
+
+        }
+    }
+
+    @Override
+    public Collection<ValidationResult> customValidate(ValidationContext context) {
+        final Collection<ValidationResult> results = super.customValidate(context);
+        int newSize = context.getProperty(PROP_MAX_QUEUE_SIZE).asInteger();
+        if (mqttQueue == null) {
+            mqttQueue = new LinkedBlockingQueue<>(context.getProperty(PROP_MAX_QUEUE_SIZE).asInteger());
+        }
+        int msgPending = mqttQueue.size();
+        if (msgPending > newSize) {
+            results.add(new ValidationResult.Builder()
+                    .valid(false)
+                    .subject("ConsumeMQTT Configuration")
+                    .explanation(String.format("%s (%d) is smaller than the number of messages pending (%d).",
+                            PROP_MAX_QUEUE_SIZE.getDisplayName(), newSize, msgPending))
+                    .build());
+        }
+
+        return results;
+    }
+
+    @Override
+    protected void init(final ProcessorInitializationContext context) {
+        logger = getLogger();
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return relationships;
+    }
+
+    @Override
+    public final List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return descriptors;
+    }
+
+    @OnScheduled
+    public void onScheduled(final ProcessContext context) throws IOException, ClassNotFoundException {
+        qos = context.getProperty(PROP_QOS).asInteger();
+        maxQueueSize = context.getProperty(PROP_MAX_QUEUE_SIZE).asLong();
+        topicFilter = context.getProperty(PROP_TOPIC_FILTER).getValue();
+
+        buildClient(context);
+    }
+
+    @OnUnscheduled
+    public void onUnscheduled(final ProcessContext context) {
+        try {
+            synchronized (mqttClientConnectLock) {
+                if(mqttClient != null && mqttClient.isConnected()) {
+                    mqttClient.disconnect(DISCONNECT_TIMEOUT);
+                    logger.info("Disconnected the MQTT client.");
+                }
+            }
+        } catch(MqttException me) {
+            logger.error("Failed when disconnecting the MQTT client.", me);
+        }
+    }
+
+
+    @OnStopped
+    public void onStopped(final ProcessContext context) throws IOException {
+        if(processSessionFactory != null) {
+            logger.info("Finishing processing leftover messages");
+            ProcessSession session = processSessionFactory.createSession();
+            transferQueue(session);
+        } else {
+            if (mqttQueue!= null && mqttQueue.size() > 0){
+                throw new ProcessException("Attempting to stop processor but stored ProcessSessionFactory is not set, there are messages in the MQTT Queue and it is configured to " +
+                        "finish processing the messages.");
+            }
```

Comment:
"There still is the chance the processor would finish draining the queue" - indeed there is and that is why I pointed out to that abstract Kafka processor. I know it can't be applied one-for-one at its current state, but some aspects of it can. For example, it tracks how many tasks are currently executing. This accomplishes two things:
1. The reconnect won't happen unless all tasks completed
2. It has _acceptTasks_ flag to control the acceptance of new tasks to disallow new tasks until "some condition" resets the flag which allows for a clean resets of connection and other resources. This second point also brings an interesting point in your current code. If there are multiple tasks executing onTrigger and at some point one resets the connection (regardless of the reason) while other referenced previous instance of such connection?

By knowing/controlling the above you would always know that there are live tasks which will essentially drain the queue and if there are not then you would drain it as part of the graceful stop.